modules/overrides/user_config_overrides.go

```diff
@@ -0,0 +1,384 @@
+	if err := services.StartManagerAndAwaitHealthy(ctx, o.subservices); err != nil {
+		return errors.Wrap(err, "unable to start overrides subservices")
+	}
+
+	return o.reloadAllTenantLimits(ctx)
+}
+
+func (o *userConfigOverridesManager) loop(ctx context.Context) error {
+	ticker := time.NewTicker(o.cfg.PollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-ticker.C:
+			err := o.reloadAllTenantLimits(ctx)
+			if err != nil {
+				level.Error(o.logger).Log("msg", "failed to refresh user-configurable config", "err", err)
+			}
+			continue
+
+		case err := <-o.subservicesWatcher.Chan():
+			return errors.Wrap(err, "overrides subservice failed")
+		}
+	}
+}
+
+func (o *userConfigOverridesManager) stopping(error) error {
+	return services.StopManagerAndAwaitStopped(context.Background(), o.subservices)
+}
+
+func (o *userConfigOverridesManager) reloadAllTenantLimits(ctx context.Context) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "overrides.reloadAllTenantLimits")
+	defer span.Finish()
+
+	traceID, _ := tracing.ExtractTraceID(ctx)
+	level.Info(o.logger).Log("msg", "reloading all tenant limits", "traceID", traceID)
+
+	// TODO can we make this lock smaller? this will block all operations that read from the overrides
+	o.mtx.Lock()
+	defer o.mtx.Unlock()
+
+	// List tenants with user-configurable overrides
+	tenants, err := o.listTenantsUnderLock(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Clean up cached tenants that have been removed from the backend
+	for cachedTenant := range o.tenantLimits {
+		if doesNotContain(tenants, cachedTenant) {
+			delete(o.tenantLimits, cachedTenant)
+		}
+	}
+
+	// For every tenant with user-configurable overrides, download and cache them
+	for _, tenant := range tenants {
+		_, err = o.getTenantLimitsUnderLock(ctx, tenant)
+		if err != nil {
+			// TODO should we keep trying the other tenants and return a combined error message?
+			//  this implementation gives up after a single failure
+			return errors.Wrap(err, "failed to load tenant limits")
+		}
+	}
+
+	return nil
+}
+
+func (o *userConfigOverridesManager) listTenantsUnderLock(ctx context.Context) ([]string, error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "overrides.listTenantsUnderLock")
+	defer span.Finish()
+
+	// TODO to avoid polling the entire bucket, use a tenant list and keep it in a shared cache
+	tenants, err := o.r.List(ctx, []string{overridesKeyPath})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list tenants")
+	}
+	return tenants, nil
+}
+
+// getTenantLimits will look up the UserConfigurableLimits for a tenant, it performs a backend request
+// and will update the entry in the local cache.
+func (o *userConfigOverridesManager) getTenantLimits(ctx context.Context, userID string) (tenantLimits *UserConfigurableLimits, err error) {
+	o.mtx.Lock()
+	defer o.mtx.Unlock()
+
+	return o.getTenantLimitsUnderLock(ctx, userID)
```

Comment:
i find it a little strange (but not wrong) that calling GET to get the limits also updates the local, in memory copy.

i feel like the API should be separate from the mechanism that maintains limits internally. so an api that's just a passthrough to object storage and a separate module that polls tenant limits and provides them as part of the overrides chain.

wdyt? is this separation of concerns worth it?

Reply:
I've linked them with the idea "we have spent work to fetch the data anyway, let's immediately update our local copy". Advantage is you have a quicker update in the instance handling the request. In the context of a distributed system that isn't really a big deal though since all the other components will have to fetch it too...

I'll give separating the two a try and see if it cleans up the code a bit 🙂

Reply:
I've split up the manager, api and client structures. Agree each is simpler now and can focus on its own thing :+1:

So the manager and the API both use client to talk with the backend. All components will run the manager but only the query-frontend needs to run the API.